src/slurm_plugin/clustermgtd.py

```diff
@@ -0,0 +1,624 @@
+            raise
+
+    def _clean_up_inactive_partition(self, inactive_nodes):
+        """
+        Terminate all other instances associated with nodes directly through EC2.
+
+        This operation will not update node states, nodes will retain their states
+        Nodes will be placed into down* by slurm after SlurmdTimeout
+        If dynamic, nodes will be power_saved after SuspendTime
+        describe_instances call is made with filter on private IPs
+        """
+        try:
+            log.info("Clean up instances associated with nodes in INACTIVE partitions: %s", inactive_nodes)
+            self.instance_manager.terminate_associated_instances(
+                inactive_nodes, terminate_batch_size=self.sync_config.terminate_max_batch_size
+            )
+        except Exception as e:
+            log.error(
+                "Unable to clean up nodes in INACTIVE partitions with exception: %s\nContinuing other operations...", e
+            )
+
+    def _get_ec2_instances(self):
+        """
+        Get EC2 instance by describe_instances API.
+
+        Call is made by filtering on tags and includes non-terminating instances only
+        Instances returned will not contain instances previously terminated in _clean_up_inactive_partition
+        """
+        try:
+            return self.instance_manager.get_cluster_instances(include_master=False, alive_states_only=True)
+        except Exception as e:
+            log.error("Unable to retrieve instances from EC2 with exception: %s", e)
+            raise
+
+    def _perform_health_check_actions(self, cluster_instances, ip_to_slurm_node_map):
+        """Run health check actions."""
+        try:
+            id_to_instance_map = {instance.id: instance for instance in cluster_instances}
+            # Get instance health states
+            instance_health_states = self.instance_manager.get_instance_health_states(list(id_to_instance_map.keys()))
+            # Perform EC2 health check actions
+            if not self.sync_config.disable_ec2_health_check:
+                self._handle_health_check(
+                    instance_health_states,
+                    id_to_instance_map,
+                    ip_to_slurm_node_map,
+                    health_check_type=ClusterManager.HealthCheckTypes.ec2_health,
+                )
+            # Perform scheduled event actions
+            if not self.sync_config.disable_scheduled_event_health_check:
+                self._handle_health_check(
+                    instance_health_states,
+                    id_to_instance_map,
+                    ip_to_slurm_node_map,
+                    health_check_type=ClusterManager.HealthCheckTypes.scheduled_event,
+                )
+        except Exception as e:
+            log.error(
+                "Unable to perform instance health check actions with exception: %s\nContinuing other operations...", e
```

Comment:
one of the downsides of catching the exception only here is that it's hard for us to understand where it generated from (we don't have a stack trace). If we want to add more granular logging we could use a decorator like the following (probably it's incomplete but it's enough to give the idea):

```
def log_exception(logger, desc, log_level=logging.ERROR, exception=Exception):
    def _log_exception(function):
        @functools.wraps(function)
        def wrapper(*args, **kwargs):
            try:
                return function(*args, **kwargs)
            except exception as e:
                logger.log(log_level, "Failed when %s with exception %s", desc, e)
                raise
        return wrapper

    return _log_exception


@log_exception(logger, "retrieving EC2 data")
def my function(...):
  ...
```

Reply:
Will look through this